### PR TITLE
Inactive products with qty in warehouse

### PR DESCRIPTION
**Pull request by dm-ps**
refs: https://github.com/metasfresh/metasfresh/issues/18254

### CODE REVIEW

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/IHUQueryBuilder.java [hunk 1]
```diff
@@ -476,5 +476,9 @@ public interface IHUQueryBuilder
 	 */
 	IHUQueryBuilder setOnlyStockedProducts(final boolean onlyStockedProducts);
 
-
+	/**
+	 * Do not consider {@link I_M_HU#COLUMNNAME_C_BPartner_ID} in the query, so we do not filter based on this column at all
+	 * Otherwise, if {@link #addOnlyInBPartnerId(BPartnerId)} gets <code>null</code> we will search only for those records with {@link I_M_HU#COLUMNNAME_C_BPartner_ID} <code>null</code>.
+	 */
+	IHUQueryBuilder setIgnoreBPartnerColumn();
```
Comment (teosarca) [inline backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/IHUQueryBuilder.java:483]: why do we need this flag?
precisely which is the case we cannot achieve by onlyIfAssignedToBPartner and OnlyInBPartnerIds fields.

PS: I have the feeling we are overcomplicating things for future devs.

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/model/validator/M_Product.java [hunk 2]
```diff
@@ -58,4 +68,31 @@ public void updateAllM_HU_PI_Item_Products(final I_M_Product product)
...
+		final IStorageQuery storageQuery = storageEngine
+				.newStorageQuery()
+				.addProductId(ProductId.ofRepoId(product.getM_Product_ID()))
+				.setExcludeAfterPickingLocator(false)
+				.setOnlyStockedProducts(true)
+				.setOnlyActiveHUs(true)
+				.setIgnoreBPartnerColumn();
+
+		final IContextAware context = PlainContextAware.createUsingOutOfTransaction();
+		final boolean isEmptyStock = storageEngine.retrieveStorageRecords(context, storageQuery).isEmpty();
```
Comment (teosarca) [inline backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/model/validator/M_Product.java:91]: why not introducing storageEngine.anyMatching(storageQuery) method?

i.e. why shall we load all records when we just want to check if there is any matching.

File: backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/material/cockpit/filters/ProductFilterUtil.java [hunk 3]
```diff
@@ -89,6 +87,18 @@ public static DocumentFilterDescriptor createFilterDescriptor()
 				.setWidgetType(DocumentFieldWidgetType.YesNo)
 				.setOperator(Operator.EQUAL);
 
+		final DocumentFilterParamDescriptor.Builder isActive = DocumentFilterParamDescriptor.builder()
+				.setFieldName(I_M_Product.COLUMNNAME_IsActive)
+				.setDisplayName(Services.get(IMsgBL.class).translatable(I_M_Product.COLUMNNAME_IsActive))
```
Comment (teosarca) [inline backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/material/cockpit/filters/ProductFilterUtil.java:92]: 1. either extract `Services.get(IMsgBL.class)` as a local var and use it in all 3 cases u have,
2. or use de.metas.i18n.TranslatableStrings#adElementOrMessage

**Review by teosarca — CHANGES_REQUESTED** [review]
check my comments

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/IHUQueryBuilder.java (hunk 1 above)
Comment (dm-ps) [reply backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/IHUQueryBuilder.java:483]: When we have a Picked HU for which the BPartner is set, but I don't know that BPartner directly from the product (interceptor), so I want to entirely omit `M_HU.C_BPartner_ID` when fetching the records.
By default in `de.metas.storage.spi.hu.impl.HUStorageQuery#HUStorageQuery`, `OnlyIfAssignedToBPartner` is set to `false` and `addOnlyInBPartnerId` to null.
Therefore, in `de.metas.handlingunits.impl.HUQueryBuilder#createQueryFilter` we'll end up with a Set which is not empty, and we'll filter only for those record with `C_BPartner_ID IS NULL`.
image

File: backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/model/validator/M_Product.java (hunk 2 above)
Comment (dm-ps) [reply backend/de.metas.handlingunits.base/src/main/java/de/metas/handlingunits/model/validator/M_Product.java:91]: will do, thx

File: backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/material/cockpit/filters/ProductFilterUtil.java (hunk 3 above)
Comment (dm-ps) [reply backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/material/cockpit/filters/ProductFilterUtil.java:92]: will do the first option, thx

**Review by teosarca — APPROVED** [review]

**Pull request merged**